### PR TITLE
Add remaining runtime tests

### PR DESCRIPTION
**Pull request by khansaad**
## Description

This PR updates the generate_recommendations test to include more runtime recommendations scenarios. 

- Adds a separate util file for runtime

Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] New feature
- [ ] Docs update
- [ ] Breaking change (What changes might users need to make in their application due to this PR?)
- [ ] Requires DB changes

## How has this been tested?

Please describe the tests that were run to verify your changes and steps to reproduce. Please specify any test configuration required. 

- [x] New Test X
- [x] Functional testsuite

**Test Configuration**
* Kubernetes clusters tested on: 

## Checklist :dart:

- [ ] Followed coding guidelines
- [ ] Comments added
- [ ] Dependent changes merged
- [ ] Documentation updated
- [ ] Tests added or updated

## Additional information

Include any additional information such as links, test results, screenshots here

## Summary by Sourcery

Expand runtime recommendations coverage in local monitoring tests by adding scenario-specific test cases and extracting shared runtime helpers into a dedicated utility module.

Enhancements:
- Introduce a reusable runtime_utils helper module for generating recommendations, validating runtime-related env vars, and inspecting GC flag patterns in recommendation outputs.
- Refactor existing runtime recommendation test to use the new shared helper for generating and validating recommendations.
- Centralize runtime-related constants and utilities previously in generic helpers into the new runtime utilities module.

Tests:
- Add multiple runtime-focused tests covering Semeru GC flags, missing JVM runtime metadata, JVM version edge cases, layer/runtime mismatches, and non-runtime-capable datasources.

### CODE REVIEW

File: tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py [hunk 1]
```diff
@@ -47,110 +49,184 @@ def test_runtime_recommendation(cluster_type):
...
+    if not _contains_any_pattern(env_values, SEMERU_GC_PATTERNS):
+        pytest.skip("Semeru/OpenJ9 GC policy not detected for current workload – skipping Semeru-specific assertion")
 
-    response = create_metadata_profile(metadata_profile_json_file)
-    data = response.json()
-    print(data["message"])
-    assert response.status_code == SUCCESS_STATUS_CODE
-    assert data["status"] == SUCCESS_STATUS
+    assert _contains_any_pattern(env_values, SEMERU_GC_PATTERNS), (
+        f"Expected Semeru GC policy flags {SEMERU_GC_PATTERNS} in JAVA_OPTIONS, got: {env_values}"
```
[the remark this answers is not in the document]
Comment (khansaad) [reply tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py]: Updated now

File: tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py [hunk 2]
```diff
@@ -47,110 +49,184 @@ def test_runtime_recommendation(cluster_type):
...
+def test_non_runtime_supported_datasource_logs_message_on_generate(cluster_type):
+    """
+    Test Description:
+    For datasources that exist but do NOT support runtime recommendations, generateRecommendations
+    should still succeed but the server should log RUNTIMES_RECOMMENDATIONS_NOT_AVAILABLE.
+    """
+    input_json_file = "../json_files/create_tfb_exp.json"
+    form_kruize_url(cluster_type)
+
+    # Use bulk metadata profile that defines datasources with and without runtime support
```
[the remark this answers is not in the document]
Comment (khansaad) [reply tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py]: This was a redundant test, we already have one at the create_experiment level. Removed it now

**Comment by chandrams** [comment]
@khansaad Have you run these tests? I see these failures - https://ci.app-svc-perf.corp.redhat.com/job/ExternalTeams/job/Autotune/job/kruize_functional_tests/447/

**Comment by kusumachalasani** [comment]
@khansaad Can you please post the latest results ?

File: tests/scripts/common/common_functions.sh [hunk 3]
```diff
@@ -1021,3 +1021,49 @@ function kruize_local_patch() {
...
+###########################################
+function benchmarks_install() {
+	APP_NAMESPACE="default"
+
+	echo
+	echo "#######################################"
+	pushd benchmarks >/dev/null
+    echo "5. Installing TechEmpower (Quarkus REST EASY) benchmark into cluster"
+    pushd techempower >/dev/null
+    kubectl apply -f manifests/default_manifests -n ${APP_NAMESPACE}
```
Comment (kusumachalasani) [inline tests/scripts/common/common_functions.sh]: For runtimes specific scripts, we want to use manifests from kruize-demos - which triggers the load by default without an external command which is as part of - https://github.com/kruize/kruize-demos/pull/177

**Comment by khansaad** [comment]
> @khansaad Can you please post the latest results ?
[runtimes-test-results.zip](https://github.com/user-attachments/files/26495280/runtimes-test-results.zip)

**Comment by chandrams** [comment]
@khansaad Why is semeru test skipped here? Can you submit it on jenkins as in the past we have seen failures on jenkins

File: tests/scripts/common/common_functions.sh (hunk 3 above)
Comment (chandrams) [reply tests/scripts/common/common_functions.sh]: @khansaad Can you address the above comment

File: tests/scripts/helpers/runtime_utils.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
+"""
+Copyright (c) 2022, 2024 Red Hat, IBM Corporation and others.
```
Comment (chandrams) [inline tests/scripts/helpers/runtime_utils.py]: Update the copyright year

File: tests/scripts/helpers/runtime_utils.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+                        if isinstance(env_list, list):
+                            envs.extend(env_list)
+    return envs
+
+
+def _env_values(list_reco_json):
+    return [env.get("value", "") for env in _extract_runtime_envs(list_reco_json)]
+
+
+def _contains_any_pattern(values, patterns):
```
Comment (chandrams) [inline tests/scripts/helpers/runtime_utils.py:764]: Please add description for the functions

File: tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py [hunk 6]
```diff
@@ -47,110 +51,154 @@ def test_runtime_recommendation(cluster_type):
...
-    # Remove benchmarks directory
-    shutil.rmtree("benchmarks")
 
-    # Clean up temp experiment JSON file
-    if temp_input_json_file and os.path.exists(temp_input_json_file):
-        os.unlink(temp_input_json_file)
+@pytest.mark.runtimes
+def test_no_gc_recommendation_when_jvm_version_missing(cluster_type):
+    """
+    Test Description: If jvm_info metrics are present but the version label is not part
```
Comment (chandrams) [inline tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py:127]: Update the test docs with these scenarios

File: tests/scripts/helpers/runtime_utils.py (hunk 4 above)
Comment (khansaad) [reply tests/scripts/helpers/runtime_utils.py]: Done

File: tests/scripts/helpers/runtime_utils.py (hunk 5 above)
Comment (khansaad) [reply tests/scripts/helpers/runtime_utils.py:764]: Added

File: tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py (hunk 6 above)
Comment (khansaad) [reply tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py:127]: Updated

**Comment by chandrams** [comment]
@kusumachalasani Are you fine with the changes?

**Comment by kusumachalasani** [comment]
> @kusumachalasani Are you fine with the changes?

I don't see the changes related to manifests using from "manifests/kruize-demos" while deploying benchmarks.

**Comment by khansaad** [comment]
> > @kusumachalasani Are you fine with the changes?
> 
> I don't see the changes related to manifests using from "manifests/kruize-demos" while deploying benchmarks.

Added now

File: tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py [hunk 7]
```diff
@@ -124,6 +116,7 @@ def remove_jvm_metrics(metric_profile_json):
     )
 
 
+@pytest.mark.skip(reason="This will be enabled once the conditional check PR is merged(PR #1866)")
 @pytest.mark.runtimes
 def test_no_gc_recommendation_when_jvm_version_missing(cluster_type):
```
Comment (kusumachalasani) [inline tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py]: Please make this testcase as positive as the output you are getting as expected and is a default behaviour.

Comment (khansaad) [reply tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py]: Done

**Comment by khansaad** [comment]
@chandrams @kusumachalasani Able to see it running both on minikube and openshift
https://ci.app-svc-perf.corp.redhat.com/job/ExternalTeams/job/Autotune/job/kruize_functional_tests/466
https://ci.app-svc-perf.corp.redhat.com/job/ExternalTeams/job/Autotune/job/kruize_functional_tests/465

**Comment by chandrams** [comment]
@khansaad I see unexpected test results, but the tests till seem to pass. Can you check & fix these.

test_runtime_recommendation - ENV doesn't have quarkus recommendations, but the test passes (this could be due to setup, but test has to fail when quarkus reco is not present)

test_no_recommendation_for_layer_runtime_mismatch - Heap & GC recommendations are seen, but the test passes

File: tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py [hunk 8]
```diff
@@ -47,110 +55,168 @@ def test_runtime_recommendation(cluster_type):
...
+    )
+
+    env_values = _env_values(list_reco_json)
+    assert _contains_any_pattern(env_values, HOTSPOT_GC_PATTERNS + SEMERU_GC_PATTERNS), (
+        f"GC-related runtime env are expected when JVM metrics are missing, but not found"
+    )
+
+
+@pytest.mark.runtimes
+def test_no_gc_recommendation_when_jvm_version_missing(cluster_type):
```
Comment (chandrams) [inline tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py]: Rename the test as no_gc_recommendation is misleading

File: tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py [hunk 9]
```diff
@@ -47,110 +55,168 @@ def test_runtime_recommendation(cluster_type):
...
+        env for env in semeru_envs if env.get("name") not in (JAVA_OPTIONS, JDK_JAVA_OPTIONS)
+    ]
+    assert not invalid_envs, (
+        "Semeru GC policy flags were found, but not on expected env vars "
+        f"{JDK_JAVA_OPTIONS}/{JAVA_OPTIONS}. Offending entries: {invalid_envs}"
+    )
+
+
+@pytest.mark.runtimes
+def test_no_runtime_recommendations_when_jvm_metadata_missing(cluster_type):
```
Comment (chandrams) [inline tests/scripts/local_monitoring_tests/rest_apis/test_generate_recommendation.py]: Rename this test as no_runtime_recommendations is misleading

**Comment by chandrams** [comment]
> @khansaad I see unexpected test results, but the tests till seem to pass. Can you check & fix these.
> 
> test_runtime_recommendation - ENV doesn't have quarkus recommendations, but the test passes (this could be due to setup, but test has to fail when quarkus reco is not present)
> 
> test_no_recommendation_for_layer_runtime_mismatch - Heap & GC recommendations are seen, but the test passes

Can you share the latest result links

**Comment by chandrams** [comment]
@Saad I have tested your changes in PR 1824 - https://ci.app-svc-perf.corp.redhat.com/job/ExternalTeams/job/Autotune/job/kruize_functional_tests/483/.  I see 2 issues - Semeru test is skipped on minikube, runtime reco - doesn't have quarkus reco in env, are you enabling it in kube state metrics?

**Comment by chandrams** [comment]
@khansaad Please create an issue to investigate the runtimes failures on minikube.

**Comment by khansaad** [comment]
> investigate the runtimes failures on minikube

Created [1884](https://github.com/kruize/autotune/issues/1884) to track this
